saged/normalize_recount_data.py

```diff
@@ -0,0 +1,178 @@
+    ---------
+    counts: The array of transcript counts per gene
+    gene_length_arr: The array of lengths for each gene in counts
+
+    Returns
+    -------
+    tpm: The tpm normalized expression data
+    """
+    counts = np.array(counts, dtype=float)
+
+    reads_per_kb = counts / gene_length_arr
+
+    sample_total_counts = np.sum(counts)
+    per_million_transcripts = sample_total_counts / 1e6
+
+    tpm = reads_per_kb / per_million_transcripts
+
+    return tpm
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument('count_file', help='The file containing the count matrix generated by '
+                                           'download_recount3.R')
+    parser.add_argument('gene_file', help='The file with gene lengths from get_gene_lengths.R')
+    parser.add_argument('out_file', help='The file to save the normalized results to')
+    parser.add_argument('--num_genes', help='The script will keep the top K most variable genes. '
+                                            'This flag sets K.', default=5000, type=int)
+
+    args = parser.parse_args()
+
+    gene_to_len = parse_gene_lengths(args.gene_file)
+
+    with open(args.count_file, 'r') as count_file:
+        header = count_file.readline()
+        header = header.replace('"', '')
+        header_genes = header.strip().split('\t')
+        header_genes = [gene.split('.')[0] for gene in header_genes]
+
+        bad_indices = []
+        gene_length_arr = []
+        for i, gene in enumerate(header_genes):
+            if gene not in gene_to_len.keys():
+                bad_indices.append(i)
+            else:
+                gene_length_arr.append(gene_to_len[gene])
+
+        gene_length_arr = np.array(gene_length_arr)
+
+        means = None
+        M2 = None
+        maximums = None
+        minimums = None
+
+        # First time through the data, calculate statistics
+        for i, line in tqdm.tqdm(enumerate(count_file), total=317475):
+            line = line.replace('"', '')
+            line = line.strip().split('\t')
+            try:
+                # Thanks to stackoverflow for this smart optimization
+                # https://stackoverflow.com/a/11303234/10930590
+                counts = line[1:]  # bad_indices is still correct because of how R saves tables
+                for index in reversed(bad_indices):
+                    del counts[index]
+
+                tpm = calculate_tpm(counts, gene_length_arr)
+
+                if any(np.isnan(tpm)):
+                    continue
+
+                # Online variance calculation https://stackoverflow.com/a/15638726/10930590
+                if means is None:
+                    means = tpm
+                    M2 = 0
+                    maximums = tpm
+                    minimums = tpm
+                else:
+                    delta = tpm - means
+                    means = means + delta / (i + 1)
+                    M2 = M2 + delta * (tpm - means)
```

Comment:
Tiny tiny thing, but why not say `M2 = M2 + delta * delta`?

Reply:
Good question, I thought the same thing when I first implemented the calculation. Since `means` is redefined on 119, the values for `delta` and `tpm - means` on line 120 are different